pkg/crc/api/handlers.go

```diff
@@ -26,20 +31,37 @@ func NewHandler(config crcConfig.Storage, machine machine.Client) *Handler {
 			Underlying: machine,
 		},
 		Config: config,
+		lock:   semaphore.NewWeighted(int64(1)),
 	}
 }
 
+/*
+* Start, Stop and Delete  handlers are serialized with a semaphore
+* since they modify VM state  they should not be ran  concurrently
+* The other handlers viz. Status, GetVersion, GetConfig, SetConfig
+* UnsetConfig and GetWebconsoleInfo are thread safe and can be ran
+* concurrently
```

Comment:
`viz.`? dunno if it's intentional/not sure what it means?

Reply:
This statement seems strongly biased by the needs of the trays ;) If Start/Stop/Delete modify VM state, why are we allowing `Status` to run concurrently with them, as the `Status` we return is most likely going to be outdated very soon?
Running `SetConfig`/`UnsetConfig` concurrently can also be misleading, since we return the list of the keys which were successfully set/unset. `GetConfig/SetConfig/UnsetConfig` are only thread safe as long as no other thread calls `AddSetting` while they are running.

Reply:
`Status` can also be `Starting` or `Stopping` so if the status is queried when a `Start` `Stop` or `Delete` is happening concurrently it'll actually show the correct status (as long as it can fetch the status) even if it gets outdated its the status at that point of time.

> GetConfig/SetConfig/UnsetConfig are only concurrency safe as long as no other goroutine calls AddSetting while they are running.

yeah, but none of the API handlers will be calling it, we'll always be using the config object created in root.go

Reply:
> viz.? dunno if it's intentional/not sure what it means?

Its supposed to mean "namely" i used 'viz.' to keep the line of the same length as the one above it :)

Reply:
ah, very first time I see this. It's supposed to be surrounded by commas, so the line lengths won't match ;)

Reply:
>>    GetConfig/SetConfig/UnsetConfig are only concurrency safe as long as no other goroutine calls AddSetting while they are running.

> yeah, but none of the API handlers will be calling it, we'll always be using the config object created in root.go

This statement is missing a very important word, this word is "today". The code is currently written this way, and it's why it's indeed not an issue. But we now have an undocumented/unenforced requirement that no `crc` code should be calling `AddSetting` after the daemon listeners  have been started. If we forget about this, we'll start having random crashes or bugs when using the daemon.

Reply:
> If Start/Stop/Delete modify VM state, why are we allowing Status to run concurrently with them, as the Status we return is most likely going to be outdated very soon?

A status will ALWAYS reflect a moment in time, from the time it was requested. Otherwise we need to implement a streaming model... with all the complexity involved. This sounds like improvements.

Reply:
Dropped the serialization commit as PR #2101 addresses this